### ISSUE TITLE
Using google-cloud-graalvm-support with Gradle

### ISSUE THREAD

**Issue by lpellegr** [GoogleCloudPlatform/google-cloud-graalvm-support#58]
When creating a native image for a Micronaut 2 app (making use of Firebase and thus `google-cloud-firestore`) with Gradle, I am getting the following error:

```
> Task :nativeImage FAILED
Error: Invalid or corrupt jarfile /home/lpellegr/.gradle/caches/modules-2/files-2.1/org.graalvm.nativeimage/svm-hosted-native-linux-amd64/20.3.0/1eb9f7ea955f02c49cc5f3fa722f388cb15e8dd5/svm-hosted-native-linux-amd64-20.3.0.tar.gz
Caused by: java.util.zip.ZipException: zip END header not found

FAILURE: Build failed with an exception.
```

when `google-cloud-graalvm-support` is declared as a compile or compileOnly dependency:

```
dependencies {
  //  ...
  compile 'com.google.cloud:google-cloud-graalvm-support:0.1.0'
}
```

I tried removing the cached archive file but this happens again and again. This lets me think I am not using google-cloud-`graalvm-support dependency` as it should, or maybe Gradle is not supported. Do you support Gradle? how should we configure a Gradle project?

**Comment by lpellegr** [comment]
@dzou any ideas about how to use this dependency with Gradle?

**Comment by dzou** [comment]
@lpellegr - Hello! Thank you for the report, apologies for the delay as we were on holiday the last couple days.

I'll take a look at this and try to get a working sample for Gradle. Conceptually, our dependency doesn't do anything special, it just adds a couple classes and config files to the classpath for the GraalVM compiler to pick up when it runs. 

I've seen `Zip` related exceptions before and I think I may have been able to fix it by tweaking the version of GraalVM I used; i.e. switching to the latest version (upgrading from 20.1.0 to 20.2.0) resolved one for me. Not sure if it will work for you.

But regardless, let me see if I can get a gradle sample working in the next few days, will keep you posted.

**Comment by lpellegr** [comment]
Hi, @dzou (and Happy new year). Thanks for your answer. 

For your information, I am getting the error above with the most recent version of GraalVM:

```
$ java --version
openjdk 11.0.9 2020-10-20
OpenJDK Runtime Environment GraalVM CE 20.3.0 (build 11.0.9+10-jvmci-20.3-b06)
OpenJDK 64-Bit Server VM GraalVM CE 20.3.0 (build 11.0.9+10-jvmci-20.3-b06, mixed mode, sharing)
```

Looking forward to reading about your try out.

**Comment by dzou** [comment]
I see, thanks for the update!

Any chance you could share your gradle file too? Am curious to see what other deps you might have (like what version of the graalvm plugin you use and flags).

**Comment by lpellegr** [comment]
@dzou I am using Micronaut. Here is a repository to reproduce the issue. It exposes a basic endpoint that increments a counter in Firestore and returns a message with the counter value:

https://github.com/lpellegr/micronaut-graalvm-gcp

To start a basic run from a JVM:

```
$ GOOGLE_APPLICATION_CREDENTIALS=/path/to/serviceaccount.json gradle run
```

then, when you open http://localhost:8080, you see the message with the counter value.

Now, let's say you want to build a native image:

```
$ gradle nativeImage -i
```

The execution fails with the error:
> java.util.zip.ZipException: zip END header not found

Running the Gradle task with option `-i` shows the native-image command that is used. Here is the output from my side:
https://gist.githubusercontent.com/lpellegr/221245b2e4db09149cf7859fe88b3fd1/raw/d75a3b1488e868a2797297ab3c7c0244720d5f55/gistfile1.txt

If you remove the compile dependency to `com.google.cloud:google-cloud-graalvm-support:0.1.0` in build.gradle, then, the execution no longer fails due to a ZipException.

**Comment by dzou** [comment]
Ah thank you, this is really helpful. I'll take a look and followup shortly.

**Comment by dzou** [comment]
Ok, so I think I have a tentative working solution for you. You can try modifying your dependency like this for now:

```
compile('com.google.cloud:google-cloud-graalvm-support:0.1.0') {
        exclude group: "org.graalvm.nativeimage"
        exclude group: "org.graalvm.truffle"
        exclude group: "org.graalvm.sdk"
        exclude group: "org.graalvm.compiler"
}
```

It seems to work for me after this change.

Basically the GraalVM libraries we depend on in our artifact seem to pull in some transitive dependencies that breaks something in Gradle. It looks like other tools [encountered the same issue](https://youtrack.jetbrains.com/issue/KT-41313) some time ago.

I think I will embed this solution in our `0.2.0` release so you don't have to worry about it and then we can wait for a true solution later. I also followed up on your issue here: https://github.com/oracle/graal/issues/3093 , will see what their reaction is.

**Comment by lpellegr** [comment]
Thanks for the quick follow up. I confirm the workaround works also on my side.

**Comment by dzou** [comment]
@lpellegr - No problem, glad to hear it works.

Just a heads up, I realized that the root error/mistake on our end is that we actually need to mark our dependency on the GraalVM sdk libraries as `provided` (like this):

```
    <dependency>
      <groupId>org.graalvm.nativeimage</groupId>
      <artifactId>svm</artifactId>
      <version>${graalvm.version}</version>
      <scope>provided</scope>
    </dependency>
```

This seems to be the correct solution to the problem; I will submit the change and do a release for `0.2.0` soon so you don't have to rely on the workaround.

**Comment by dzou** [comment]
We just released version `0.2.0` of the project.

You can try just depending on version `0.2.0` now and omitting the exclusions, I think it will now work for you.

Going to mark this as closed for now. If you still have issues just let me know.

**Comment by lpellegr** [comment]
It is working perfectly. Thanks again!
